File: lib/model/golist_model.dart
import 'package:collection/collection.dart';
import 'package:flutter/foundation.dart';
import 'package:uuid/uuid.dart';

abstract class GoListModel {
  late String id;
  late DateTime modified;
  late bool _deleted;
  late String _name;

  GoListModel(
      {required String name, String? id, DateTime? modified, bool? deleted}) {
    this.id = id ?? const Uuid().v4();
    this.modified = modified ?? DateTime.now();
    _name = name;
    _deleted = deleted ?? false;
  }

  GoListModel.fromJson(Map<String, dynamic> json);

  Map<String, dynamic> toJson() => {
        'id': id,
        'deleted': deleted,
        'modified': modified.toIso8601String(),
        'name': name
      };

  T lastModified<T extends GoListModel>(T a, T b) {
    if (a.modified.isAfter(b.modified)) {
      return a;
    }
    return b;
  }

  bool modifiedAtLeastNDaysBefore(int days) =>
      modified.isBefore(DateTime.now().subtract(Duration(days: days)));

  @override
  bool operator ==(Object other) =>
      other is GoListModel &&
      const DeepCollectionEquality().equals(toJson(), other.toJson());

  @override
  int get hashCode => toJson().hashCode;

  static int compareByModified(GoListModel a, GoListModel b) {
    return b.modified.compareTo(a.modified);
  }

  static bool Function(GoListModel) equalsById(String id) {
    return (e) => e.id == id;
  }

  static bool Function(GoListModel) equalsByName(GoListModel other) {
    return (e) =>
        e.name.trim().toLowerCase() == other.name.trim().toLowerCase();
  }

  bool equals(GoListModel other) {
    return mapEquals(toJson(), other.toJson());
  }

  bool get deleted => _deleted;

  set deleted(bool deleted) {
    modified = DateTime.now();
    _deleted = deleted;
  }

  String get name => _name;

  set name(String name) {
    modified = DateTime.now();
    _name = name;
  }

  GoListModel merge(GoListModel other);
}

File: svelte-golist/src/lib/models/GoListModel.ts
import { v4 as uuidv4 } from 'uuid';

export abstract class GoListModel {
  id: string;
  name: string;
  deleted: boolean;
  modified: Date;

  constructor({
    id,
    name,
    deleted = false,
    modified
  }: {
    id?: string;
    name: string;
    deleted?: boolean;
    modified?: Date;
  }) {
    this.id = id || uuidv4();
    this.name = name;
    this.deleted = deleted;
    this.modified = modified || new Date();
  }

  toJson(): Record<string, any> {
    return {
      id: this.id,
      name: this.name,
      deleted: this.deleted,
      modified: this.modified.toISOString()
    };
  }

  abstract merge(other: GoListModel): GoListModel;
  abstract equals(other: GoListModel): boolean;

  protected lastModified(model1: GoListModel, model2: GoListModel): GoListModel {
    return model1.modified > model2.modified ? model1 : model2;
  }
}